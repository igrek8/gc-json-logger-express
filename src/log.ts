import { NextFunction, Request, Response } from 'express';
import { Logger, Severity } from 'gc-json-logger';
import ms = require('ms');

import { LogTransformFunction } from './types/LogTransformFunction';
import { clone } from './utils/clone';

const passThrough: LogTransformFunction = (_req, _res, entry) => entry;

export function log(transform: LogTransformFunction = passThrough) {
  return (req: Request, res: Response, next: NextFunction) => {
    const start = new Date();
    const uuid = res.getHeader('x-response-id')?.toString();
    const logger = new Logger(uuid);
    Logger.setLogger(logger);

    /**
     * Capture outgoing response body
     */
    let _body: unknown;

    const _json = res.json;
    res.json = function json(body?: unknown) {
      _body ??= body;
      res.json = _json;
      return res.json(body);
    };

    const _send = res.send;
    res.send = function send(body?: unknown) {
      _body ??= body;
      res.send = _send;
      return res.send(body);
    };

    res.once('finish', () => {
      try {
        const end = new Date();
        const latency = ms(end.getTime() - start.getTime());
        const entry = transform(req, res, {
          severity: Severity.INFO,
          message: `${res.statusCode} ${req.method} ${req.url} (${latency})`,
          meta: {
            httpRequest: {
              protocol: `${req.protocol}/${req.httpVersion}`.toUpperCase(),
              requestSize: req.socket.bytesRead.toString(),
              remoteIp: req.ip,
              requestUrl: req.protocol + '://' + req.get('host') + req.originalUrl,
              requestMethod: req.method,
              userAgent: req.header('User-Agent')?.toString(),
              serverIp: `${req.socket.localAddress}:${req.socket.localPort}`,
              referer: req.header('Referer'),
              latency,
              responseSize: req.socket.bytesWritten.toString(),
              status: res.statusCode,
            },
            httpRequestHeaders: clone(req.headers),
            httpRequestBody: clone(req.body),
            httpResponseHeaders: clone(res.getHeaders()),
            httpResponseBody: clone(_body),
          },
        });
        logger.log(entry.severity, entry.message, entry.meta);
      } catch (err) {
        /* istanbul ignore next */
      }
    });
    next();
  };
}
