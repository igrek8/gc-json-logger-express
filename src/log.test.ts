import * as express from 'express';
import { Logger } from 'gc-json-logger';
import { Server } from 'http';
import * as request from 'supertest';
import { log } from './log';

jest.useFakeTimers();
jest.setSystemTime(new Date(0));

describe('log', () => {
  let server: Server;

  beforeAll((done) => {
    server = express()
      .use(log())
      .use(express.json())
      .use(express.text())
      .post('/json', (req, res) => res.status(200).json(req.body))
      .post('/text', (req, res) => res.status(200).send(req.body))
      .listen(50000, done);
  });

  afterAll((done) => {
    server.close(done);
  });

  it('supports text endpoint', async () => {
    const write = jest.spyOn(Logger.prototype, 'log');
    write.mockImplementation(() => {});
    await request(server)
      .post('/text?key=1')
      .set('referer', 'https://example.com/')
      .set('user-agent', 'curl/7.64.1')
      .set('content-type', 'text/plain')
      .send('{ "test": 1 }');
    expect(write.mock.calls).toMatchSnapshot();
  });

  it('supports json endpoint', async () => {
    const write = jest.spyOn(Logger.prototype, 'log');
    write.mockImplementation(() => {});
    await request(server)
      .post('/json?key=1')
      .set('referer', 'https://example.com/')
      .set('user-agent', 'curl/7.64.1')
      .set('content-type', 'application/json')
      .send('{ "test": 1 }');
    expect(write.mock.calls).toMatchSnapshot();
  });
});
