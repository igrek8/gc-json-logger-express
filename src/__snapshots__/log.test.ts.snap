// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log supports json endpoint 1`] = `
[
  [
    200,
    "200 POST /json?key=1",
    {
      "httpRequest": {
        "protocol": "HTTP/1.1",
        "referer": "https://example.com/",
        "remoteIp": "::ffff:127.0.0.1",
        "requestMethod": "POST",
        "requestSize": "224",
        "requestUrl": "http://127.0.0.1:50000/json?key=1",
        "serverIp": "::ffff:127.0.0.1:50000",
        "userAgent": "curl/7.64.1",
      },
    },
  ],
  [
    200,
    "200 POST /json?key=1 (0ms)",
    {
      "httpRequest": {
        "latency": "0ms",
        "responseSize": "216",
        "status": 200,
      },
      "request": {
        "body": {
          "test": 1,
        },
        "headers": {
          "accept-encoding": "gzip, deflate",
          "connection": "close",
          "content-length": "13",
          "content-type": "application/json",
          "host": "127.0.0.1:50000",
          "referer": "https://example.com/",
          "user-agent": "curl/7.64.1",
        },
      },
      "response": {
        "body": {
          "test": 1,
        },
        "headers": {
          "content-length": "10",
          "content-type": "application/json; charset=utf-8",
          "etag": "W/"a-0IjPMzN9x5tJczAzOeVnVWcl4yo"",
          "x-powered-by": "Express",
        },
      },
    },
  ],
]
`;

exports[`log supports text endpoint 1`] = `
[
  [
    200,
    "200 POST /text?key=1",
    {
      "httpRequest": {
        "protocol": "HTTP/1.1",
        "referer": "https://example.com/",
        "remoteIp": "::ffff:127.0.0.1",
        "requestMethod": "POST",
        "requestSize": "218",
        "requestUrl": "http://127.0.0.1:50000/text?key=1",
        "serverIp": "::ffff:127.0.0.1:50000",
        "userAgent": "curl/7.64.1",
      },
    },
  ],
  [
    200,
    "200 POST /text?key=1 (0ms)",
    {
      "httpRequest": {
        "latency": "0ms",
        "responseSize": "212",
        "status": 200,
      },
      "request": {
        "body": "{ "test": 1 }",
        "headers": {
          "accept-encoding": "gzip, deflate",
          "connection": "close",
          "content-length": "13",
          "content-type": "text/plain",
          "host": "127.0.0.1:50000",
          "referer": "https://example.com/",
          "user-agent": "curl/7.64.1",
        },
      },
      "response": {
        "body": "{ "test": 1 }",
        "headers": {
          "content-length": "13",
          "content-type": "text/html; charset=utf-8",
          "etag": "W/"d-1RPzMXJoVFsid1wYU/S8C5v43Kc"",
          "x-powered-by": "Express",
        },
      },
    },
  ],
]
`;
